Romanchenko 181-1 LongContest3 A

Romanchenko opened the pull request:
https://official.contest.yandex.ru/contest/11654/run-report/19528821/

Romanchenko committed:
Create hashmap.h

Romanchenko committed:
Resize added

submit 19730150
https://official.contest.yandex.ru/contest/11654/run-report/19730150/

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+#include <utility>
+#include <vector>
+
+size_t kMaxSize = (1 << 10);
```
в файле .h не может быть глобальных переменных, только глобальные константы (но и то лучше поместить эту константу внутрь класса как static const)

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+    typedef typename std::list<HashPair>::iterator ListIterator;
+    typedef typename std::list<HashPair>::const_iterator ConstListIterator;
+
+    private:
```
Лучше сначала разместить публичные методы, а потом уже приватные методы и приватные поля

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        template<class Inputiterator>
```
InputIterator

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+        }
+
+        HashMap& operator = (const HashMap& other) {
+            HashMap tmp_hashmap(other.begin(), other.end());
```
По факту получается, что мы дважды создаём новую хештаблицу, что выглядит неоправданными расходами времени

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            : size_(0)
+            , hasher_(hash_func)
+            , borders_(kMaxSize, ListIterator(nullptr)) {
+            for (auto element : init_list) {
```
const auto ?

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            }
+            ListIterator it = borders_[key_hash];
+            while (it != data_.end() && local_hash((*it).first) == key_hash) {
+                if ((*it).first == key) {
```
(*it).first -> it->first ?

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            return end();
+        }
+
+        const_iterator find(const KeyType& key) const {
```
Избавиться от повторения кода для const и non-const как в книжке

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            if (it == ListIterator(nullptr)) {
+                return;
+            }
+            if ((*it).first == key) {
```
it->first ?

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+                return;
+            }
+            if ((*it).first == key) {
+                ListIterator it2 = it;
```
давай какое-нибудь более осмысленное название сделаем, чем it2

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,349 @@
+            iterator it = find(key);
+            if (it == end()) {
+                insert(std::make_pair(key, ValueType()));
+                it = find(key);
```
Это место можно реализовать эффективнее, если метод insert будет возвращать итератор

glebshp reviewed (changes requested):
В основном нужные мелкие стилистические правки

Romanchenko committed:
Update hashmap

Сделан приватый insert_it, который возвращает итератор, исправлен оператор присваивания, ну и другие мелкие правки.
Submit 19746564
https://official.contest.yandex.ru/contest/11654/run-report/19746564/

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,356 @@
+        private:
+            size_t size_;
+            size_t capacity_;
+            constexpr static size_t init_capacity_ = (1 << 10);
```
1. Константу лучше сделать public и вынести в начало
2. Стиль именование static const должен отличаться от обычных private полей, например kInitCapacity или INIT_CAPACITY

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,356 @@
+                data_.clear();
+                borders_.clear();
+                borders_.resize(capacity_, ListIterator(nullptr));
+                for (auto element : data_copy) {
```
const ?

glebshp replied:
даже, пожалуй, const auto&

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,356 @@
+                if (found != end()) {
+                    return found;
+                }
+                if ((size_ + 1) * 2 >= capacity_) {
```
Условие перехеширования будет лучше смотреться отдельным методом

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,356 @@
+                return *this;
+            }
+            clear();
+            for (auto element : other) {
```
const auto&, и во всём остальном коде тоже надо посмотреть, где можно сделать const auto&

glebshp commented on hashmap.h:
```diff
@@ -0,0 +1,356 @@
+
+        void clear() {
+            size_ = 0;
+            for (auto & it : data_) {
```
& не следует отбивать пробелами с двух сторон, так как он здесь не является бинарным оператором

glebshp reviewed (changes requested):
Стилистические мелочи

Romanchenko committed:
Update hashmap

Const auto& и отдельная функция для условия перехэширования.
submit 19772740
https://official.contest.yandex.ru/contest/11654/run-report/19772740/

glebshp reviewed (approved):
Странно смотрятся In и Out как имена параметром метода - не сочитается с остальными стилями. В остальном ОК
